pkg/forwarder/backoff.go

```diff
@@ -0,0 +1,105 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package forwarder
+
+import (
+	"math"
+	"math/rand"
+	"time"
+
+	log "github.com/cihub/seelog"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+)
+
+const secondsFloat = float64(time.Second)
+
+var (
+	// This controls the overlap between consecutive retry interval ranges. When
+	// set to `2`, there is a guarantee that there will be no overlap. The overlap
+	// will asymptotically approach 50% the higher the value is set.
+	minBackoffFactor float64
+
+	// This controls the rate of exponential growth. Also, you can calculate the start
+	// of the very first retry interval range by evaluating the following expression:
+	// baseBackoffTime / minBackoffFactor * 2
+	baseBackoffTime float64
+
+	// This is the maximum number of seconds to wait for a retry.
+	maxBackoffTime float64
+
+	// This controls how many retry interval ranges to step down for an endpoint
+	// upon success. Increasing this should only be considered when maxBackoffTime
+	// is particularly high or if our intake team is particularly confident.
+	recoveryInterval int
+)
+
+func init() {
```

Comment:
I don't think the config is initialized at init since we load it after importing the module. I might be wrong, have you tested it ?

Also, I think the tests should be updated to use the configuration and test the min/max limit we impose here.

Reply:
I did test it. Also, based on https://medium.com/golangspec/init-functions-in-go-eac191b3860a, if `a` imports `b`, any `init` functions in `b` will execute first.

I'll update the tests now.

Reply:
The thing is that the configuration is loaded in the start command of the agent: https://github.com/DataDog/datadog-agent/blob/master/cmd/agent/app/start.go#L125

This is executed after every `init` from every imported module are executed. In Go, `init` functions are executed at import time. I tested it locally (to be sure), and indeed if I set a value in the configuration file it's not taken into account, only the default value is available.

This is also a issue with the testing you added, you don't set any value in the `config` object: https://github.com/DataDog/datadog-agent/pull/1458/commits/5296283c575f646bfcfd746d911d7b85fbc55440#diff-8e4dcd771e37bb4ab78485e4a92a717aR25, you're just testing the default value. We can say that features from viper are working and tested. What we need to test is that the new backoff system do pull value from the configuration and discard erroneous values if any are set in the configuration.

Here an example where we test how a package correctly use the config: https://github.com/DataDog/datadog-agent/blob/master/pkg/util/common_test.go#L139

Reply:
I understand now, thanks! I've updated this PR.